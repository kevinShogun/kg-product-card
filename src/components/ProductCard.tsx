import React, { createContext } from "react";
import {
	ProductContextProps,
	ProductCardProps,
} from "../interfaces/interfaces";
import { useProduct } from "../hooks/useProduct";
import styles from "../styles/styles.module.css";

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export const ProductCard = ({
	children,
	product,
	className,
	style,
	onChange,
	value,
	initialValues
}: ProductCardProps) => {
	const { counter, increaseBy, maxCount, isMaxCountReached, reset} = useProduct({ onChange, product, value, initialValues });
	
	return (
		<Provider
			value={{
				counter,
				increaseBy,
				product,
				maxCount
			}}
		>
			<div className={`${styles.productCard}  ${className}`} style={style}>
				{	
					children({
						count: counter,
						isMaxCountReached,
						maxCount: initialValues?.maxCount,
						product,
						increaseBy,
						reset
					})
				}
			</div>
		</Provider>
	);
};
