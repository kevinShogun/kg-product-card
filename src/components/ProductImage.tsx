import React, { CSSProperties, useContext } from "react";
import { ProductContext } from "./ProductCard";
import styles from "../styles/styles.module.css";
import noImg from "../assets/no-image.jpg";

export type ImageProps = {
	img?: string;
	className?: string;
	style?: CSSProperties;
};

export const ProductImage = ({ img = "", className, style }: ImageProps) => {
	const { product } = useContext(ProductContext);
	let imgShow: string;

	if (img) {
		imgShow = img;
	} else if (product.img) {
		imgShow = product.img;
	} else {
		imgShow = noImg;
	}

	return (
		<img
			src={imgShow}
			alt="product img"
			className={`${styles.productImg} ${className}`}
			style={style}
		/>
	);
};
