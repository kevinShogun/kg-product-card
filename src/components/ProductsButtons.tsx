import React, { CSSProperties, useCallback, useContext } from "react";
import { ProductContext } from "./ProductCard";
import styles from "../styles/styles.module.css";

export interface ButtonsProps {
	className?: string;
	style?: CSSProperties;
}

export const ProductsButtons = ({ className, style }: ButtonsProps) => {
	const { counter, increaseBy, maxCount } = useContext(ProductContext);

	const isMaxReached = useCallback( () => !!maxCount && counter === maxCount, [counter, maxCount]);

	return (
		<div className={`${styles.buttonsContainer}  ${className}`} style={style}>
			<button className={styles.buttonMinus} onClick={() => increaseBy(-1)}>
				{" "}-{" "}
			</button>
			<label className={styles.countLabel}>{counter}</label>
			<button
				className={`${styles.buttonAdd} ${isMaxReached() && styles.disable}`}
				onClick={() => increaseBy(+1)}
			>
				{" "}+{" "}
			</button>
		</div>
	);
};
