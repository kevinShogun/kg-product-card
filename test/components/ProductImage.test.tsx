import React from 'react';
import renderer from 'react-test-renderer';
import { ProductImage } from '../../src/components';
import { ProductCard } from '../../src/components/ProductCard';
import { product2 } from '../data/products';

describe('ProductImage ', () => {
  test('should successfully display component, with custom image', () => {
    const wrapper = renderer.create(
      <ProductImage className='custom-class' img='https://cdn.thewirecutter.com/wp-content/media/2021/12/favoritemugs-2048px-ikea.jpg?auto=webp&quality=75&width=980&dpr=2'/>
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('should successfully display component, with custom image by ProdcutCard', () => {
    const wrapper = renderer.create(
      <ProductCard product={product2}>{() => <ProductImage />}</ProductCard>
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
