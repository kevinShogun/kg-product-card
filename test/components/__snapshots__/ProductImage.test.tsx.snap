// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImage  should successfully display component, with custom image 1`] = `
<img
  alt="product img"
  className="productImg custom-class"
  src="https://cdn.thewirecutter.com/wp-content/media/2021/12/favoritemugs-2048px-ikea.jpg?auto=webp&quality=75&width=980&dpr=2"
/>
`;

exports[`ProductImage  should successfully display component, with custom image by ProdcutCard 1`] = `
<div
  className="productCard  undefined"
>
  <img
    alt="product img"
    className="productImg undefined"
    src="https://hips.hearstapps.com/hmg-prod.s3.amazonaws.com/images/coffee-mugs-1640187565.jpg?crop=0.502xw:1.00xh;0.250xw,0&resize=640:*"
  />
</div>
`;
